src/recon_buildblock/LogcoshPrior.cxx

```diff
@@ -0,0 +1,666 @@
+
+            diagonal += current;
+            prior_Hessian_for_single_densel_cast[z+dz][y+dy][x+dx] = -current*this->penalisation_factor;
+          }
+
+      prior_Hessian_for_single_densel[z][y][x]= diagonal * this->penalisation_factor;
+    }
+
+    template <typename elemT>
+    void
+    LogcoshPrior<elemT>::parabolic_surrogate_curvature(DiscretisedDensity<3,elemT>& parabolic_surrogate_curvature,
+                                                       const DiscretisedDensity<3,elemT> &current_image_estimate)
+    {
+
+      assert( parabolic_surrogate_curvature.has_same_characteristics(current_image_estimate));
+      if (this->penalisation_factor==0)
+      {
+        parabolic_surrogate_curvature.fill(0);
+        return;
+      }
+
+
+      const DiscretisedDensityOnCartesianGrid<3,elemT>& current_image_cast =
+              dynamic_cast< const DiscretisedDensityOnCartesianGrid<3,elemT> &>(current_image_estimate);
+
+      if (weights.get_length() ==0)
+      {
+        compute_weights(weights, current_image_cast.get_grid_spacing(), this->only_2D);
+      }
+
+      const bool do_kappa = !is_null_ptr(kappa_ptr);
+
+      if (do_kappa && !kappa_ptr->has_same_characteristics(current_image_estimate))
+        error("LogcoshPrior: kappa image has not the same index range as the reconstructed image\n");
+
+      const int min_z = current_image_estimate.get_min_index();
+      const int max_z = current_image_estimate.get_max_index();
+      for (int z=min_z; z<=max_z; z++)
+      {
+        const int min_dz = max(weights.get_min_index(), min_z-z);
+        const int max_dz = min(weights.get_max_index(), max_z-z);
+
+        const int min_y = current_image_estimate[z].get_min_index();
+        const int max_y = current_image_estimate[z].get_max_index();
+
+        for (int y=min_y;y<= max_y;y++)
+        {
+          const int min_dy = max(weights[0].get_min_index(), min_y-y);
+          const int max_dy = min(weights[0].get_max_index(), max_y-y);
+
+          const int min_x = current_image_estimate[z][y].get_min_index();
+          const int max_x = current_image_estimate[z][y].get_max_index();
+          for (int x=min_x;x<= max_x;x++)
+          {
+            const int min_dx = max(weights[0][0].get_min_index(), min_x-x);
+            const int max_dx = min(weights[0][0].get_max_index(), max_x-x);
+
+            elemT gradient = 0;
+            for (int dz=min_dz;dz<=max_dz;++dz)
+              for (int dy=min_dy;dy<=max_dy;++dy)
+                for (int dx=min_dx;dx<=max_dx;++dx)
+                {
+                  // psi'(t)/t = tanh/t
+                  elemT current;
+                  elemT temp = fabs(current_image_estimate[z][y][x] - current_image_estimate[z+dz][y+dy][x+dx]);
+
+                  //if (temp <= 0.01)
+                  //    current = weights[dz][dy][dx];
+                  //else
+                  current = weights[dz][dy][dx] * (1/(this->scalar))*tanh(this->scalar*(temp+0.01))/(temp+0.01);
```

Comment:
Remove `+0.01` and replace with taylor (https://github.com/UCL/STIR/pull/716#discussion_r505725258).